### ISSUE TITLE
Controller not picking up configmap updates when running with `--managed-namespace`

### ISSUE THREAD

**Issue by toblich** [argoproj/argo-workflows#11463]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

# Description

When updating the workflow controller configmap, the controller does not automatically pick up those changes. Only after I delete the controller pod and it gets recreated does it pick them up. This happens only when the controller is running with `--namespaced` and with `--managed-namespace=<other>`

For added context, the docs suggest configurations like these, with both flags, when restricting the namespace where the workflows run to just one (or a few): https://github.com/argoproj/argo-workflows/blob/2d20532a814a654075764b5901625083b690c8db/docs/managed-namespace.md?plain=1#L8-L24

# How to reproduce
To reproduce the issue:
1. Have a k8s cluster with 2 namespaces, e.g. `argo` (where the controller will run) and `argo-runs` (where workflows would run),
2. Run the workflow controller in `argo` with a workflow-controller-configmap in `argo` as well. Run it with both `--namespaced` and with `--managed-namespace=argo-runs`.
3. Edit the configmap, for example adding `workflowDefaults.spec.podMetadata.annotations: {some: value}`
4. Check the controller logs. ➡️ You won't see any log of it picking up the updated configmap, which is wrong. You'll only see the initial configuration picked up on the controller startup.
5. _(Optional)_ Run any workflow in the `argo-runs` namespace. Then, check the resulting workflow resource for it. ➡️ If you added a podMetadata annotation in step 3, you won't see it appear on the new workflow, which is wrong
6. Restart the controller (e.g. by bouncing the pod)
7. Re-do step 4 - Check the logs again. ➡️ Now, the logs will show the new configuration, as it is correctly read on startup
8. _(Optional)_ Re-do step 5 - Deploy any workflow in the `argo-runs` namespace ➡️ Now the resulting workflow resource will show the `podMetadata.annotations.some: value` defined

# Analysis

The issue comes from the following line: 
https://github.com/argoproj/argo-workflows/blob/2d20532a814a654075764b5901625083b690c8db/workflow/controller/controller.go#L389-L391 

In it, the controller is watching the config map in the **managed namespace** instead of the namespace where the controller itself is running. When it starts, and what makes most sense IMHO, it reads it from the **namespace where the controller itself runs** (as the controller AFAIK does not accept per-namespace config maps): https://github.com/argoproj/argo-workflows/blob/2d20532a814a654075764b5901625083b690c8db/workflow/controller/controller.go#L178

I've tested changing locally the namespace where it watches for the configmap and that seems to fix the issue. I'd like to further test this and make an actual PR to fix the issue myself, but I'd really appreciate any feedback/input as would be my first time contributing to this project 🙏 

# Appendix / extra info 

For reference, here is my full workflow controller deployment. See the args, where both `--namespaced` (deployed in the `argo` namespace) and `--managed-namespace=argo-runs` are specified.
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app.kubernetes.io/part-of":"argo"},"name":"workflow-controller","namespace":"argo"},"spec":{"replicas":0,"selector":{"matchLabels":{"app":"workflow-controller","app.kubernetes.io/part-of":"argo"}},"template":{"metadata":{"labels":{"app":"workflow-controller","app.kubernetes.io/part-of":"argo"}},"spec":{"containers":[{"args":["--namespaced"],"command":["workflow-controller"],"env":[{"name":"LEADER_ELECTION_IDENTITY","valueFrom":{"fieldRef":{"apiVersion":"v1","fieldPath":"metadata.name"}}}],"image":"quay.io/argoproj/workflow-controller:latest","imagePullPolicy":"Never","livenessProbe":{"failureThreshold":3,"httpGet":{"path":"/healthz","port":6060},"initialDelaySeconds":90,"periodSeconds":60,"timeoutSeconds":30},"name":"workflow-controller","ports":[{"containerPort":9090,"name":"metrics"},{"containerPort":6060}],"securityContext":{"allowPrivilegeEscalation":false,"capabilities":{"drop":["ALL"]},"readOnlyRootFilesystem":true,"runAsNonRoot":true}}],"nodeSelector":{"kubernetes.io/os":"linux"},"priorityClassName":"workflow-controller","securityContext":{"runAsNonRoot":true},"serviceAccountName":"argo"}}}}
  creationTimestamp: "2023-07-27T20:00:05Z"
  generation: 5
  labels:
    app.kubernetes.io/part-of: argo
  name: workflow-controller
  namespace: argo
  resourceVersion: "1619"
  uid: 613ec8e2-7e84-4ede-b0d8-29cf120fba08
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: workflow-controller
      app.kubernetes.io/part-of: argo
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: workflow-controller
        app.kubernetes.io/part-of: argo
    spec:
      containers:
      - args:
        - --namespaced
        - --managed-namespace=argo-runs
        command:
        - workflow-controller
        env:
        - name: LEADER_ELECTION_IDENTITY
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: quay.io/argoproj/workflow-controller:latest
        imagePullPolicy: Never
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 6060
            scheme: HTTP
          initialDelaySeconds: 90
          periodSeconds: 60
          successThreshold: 1
          timeoutSeconds: 30
        name: workflow-controller
        ports:
        - containerPort: 9090
          name: metrics
          protocol: TCP
        - containerPort: 6060
          protocol: TCP
        resources: {}
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          readOnlyRootFilesystem: true
          runAsNonRoot: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      nodeSelector:
        kubernetes.io/os: linux
      priorityClassName: workflow-controller
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        runAsNonRoot: true
      serviceAccount: argo
      serviceAccountName: argo
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2023-07-27T20:00:05Z"
    lastUpdateTime: "2023-07-27T20:13:29Z"
    message: ReplicaSet "workflow-controller-5d78457cbf" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2023-07-27T20:22:39Z"
    lastUpdateTime: "2023-07-27T20:22:39Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 5
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

```

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
<N/A>: No workflow is needed to reproduce this issue
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
<N/A>: As no workflow is needed for this issue, no wait container logs are needed either
```

**Comment by toblich** [comment]
Just for reference, there are other issues (old closed ones and other that are still open) regarding the controller picking up the updated configmap automatically. They don't specify whether they are using the `--managed-namespace` flag, so I'm not sure whether they are actually the same issue or not, but here are some of them anyways: 
- https://github.com/argoproj/argo-workflows/issues/9123 tries to propose mounting the confimap as a volume for it to be auto-refreshed. The controller already has this functionality, but as my new issue explains, it does not work when `--managed-namespace` is used
- https://github.com/argoproj/argo-workflows/issues/10187 has a similar issue, and someone proposed adding another component for this to work.
- https://github.com/argoproj/argo-workflows/issues/11194 seems similar to this issue, though mentioning only a more specific symptom

**Comment by agilgur5** [comment]
> (as the controller AFAIK does not accept per-namespace config maps):

I believe [mutex and semaphore](https://argoproj.github.io/argo-workflows/synchronization/) ConfigMaps are located in the "runs" namespace. [Artifact repositories](https://argoproj.github.io/argo-workflows/configure-artifact-repository/) can also be configured per namespace (Kubeflow Pipelines uses that functionality quite heavily).

But yes the Controller's _own_ ConfigMap is supposed to be a singular global.

> I'd like to further test this and make an actual PR to fix the issue myself, but I'd really appreciate any feedback/input as would be my first time contributing to this project 🙏

If this only affects the Controller's own ConfigMap, go for it and make the PR! You'll usually get more attention that way from maintainers, more generally speaking.

**Comment by Joibel** [comment]
This was fixed by #14675 (merged 2025-08-05), which changed the controller's own ConfigMap watcher to use the controller's installation namespace instead of the managed namespace — exactly the fix described in the analysis above. It shipped in v3.7.1 and v3.6.11.

The PR referenced the duplicate report #14673 rather than this issue, which is why this one stayed open. On current main the watcher is an informer (`newControllerConfigMapInformer`) that lists/watches `configController.GetNamespace()` with a name field selector, so the behaviour is correct there too.
